core/src/main/java/org/apache/iceberg/util/SortOrderUtil.java

```diff
@@ -67,16 +66,16 @@ public static SortOrder buildSortOrder(Schema schema, PartitionSpec spec, SortOr
 
     // make a map of the partition fields that need to be included in the clustering produced by the
     // sort order
-    Map<Pair<Transform<?, ?>, Integer>, PartitionField> requiredClusteringFields =
+    Map<Pair<String, Integer>, PartitionField> requiredClusteringFields =
         requiredClusteringFields(spec);
 
     // remove any partition fields that are clustered by the sort order by iterating over a prefix
     // in the sort order.
     // this will stop when a non-partition field is found, or when the sort field only satisfies the
     // partition field.
     for (SortField sortField : sortOrder.fields()) {
-      Pair<Transform<?, ?>, Integer> sourceAndTransform =
-          Pair.of(sortField.transform(), sortField.sourceId());
+      Pair<String, Integer> sourceAndTransform =
+          Pair.of(sortField.transform().dedupName(), sortField.sourceId());
```

Comment:
I think this needs to be `toString` rather than `dedupName`. The `dedupName` method should not really be public and shouldn't be used for cases like this. We want the actual transform here.

Reply:
Got it, I've updated the PR! 👍🏻